torch/csrc/autograd/autograd.cpp

```diff
@@ -0,0 +1,129 @@
+                              bool create_graph,
+                              at::TensorList inputs) {
+    size_t num_tensors = outputs.size();
+    size_t num_gradients = grad_outputs.size();
+    TORCH_CHECK(num_tensors == num_gradients, "got %ld tensors and %ld "
+                "gradients", num_tensors, num_gradients);
+
+    edge_list roots;
+    roots.reserve(num_tensors);
+    variable_list grads;
+    grads.reserve(num_tensors);
+
+    for (size_t i = 0; i < num_tensors; i++) {
+        at::Tensor output = outputs[i];
+        TORCH_CHECK(output.is_variable(), "element %d of tensors "
+            "tuple is not a Tensor", i);
+        const Variable& variable = as_variable_ref(output);
+        auto gradient_edge = variable.gradient_edge();
+        TORCH_CHECK(gradient_edge.function,
+            "element %d of tensors does not require grad and does not have a grad_fn", i);
+        roots.push_back(std::move(gradient_edge));
+
+        at::Tensor grad = grad_outputs[i];
+        if (!grad.defined()) {
+            // if grad is undefined
+            if (output.requires_grad()) {
+                TORCH_CHECK(output.numel() == 1, "grad can be implicitly created only for scalar outputs");
+                // create the grad with ones tensor for output that requires grad
+                grads.push_back(at::ones_like(output));
+            }
+            // otherwise we don't put grad into the grads list
+        } else if (grad.is_variable()) {
+            // if grad is a typical variable
+            // grads.push_back(as_variable_ref(grad));
+            grads.push_back(as_variable_ref(grad));
+            std::cout<<"grad :" << grads[0] << std::endl;
+        } else {
+            AT_ERROR("element %d of gradients tuple is not a Tensor or None", i);
+        }
+    }
+
+    edge_list output_edges;
+    if (inputs.empty()) {
+        size_t num_inputs = inputs.size();
+        output_edges.reserve(num_inputs);
+        for (size_t i = 0; i < num_inputs; ++ i) {
+            const Variable& input = as_variable_ref(inputs[i]);
+            const auto output_nr = input.output_nr();
+            auto grad_fn = input.grad_fn();
+            if (!grad_fn) {
+                grad_fn = input.try_get_grad_accumulator();
+            }
+            TORCH_CHECK(input.requires_grad(),
+                "One of the differentiated Tensors does not require grad");
+            if (!grad_fn) {
+                output_edges.emplace_back();
+            } else {
+                output_edges.emplace_back(grad_fn, output_nr);
+            }
+        }
+    }
+    std::cout<<"roots size: " << roots.size() << " grads size: " << grads.size() << std::endl;
+
+    return Engine::get_default_engine().execute(roots, grads, keep_graph, create_graph, output_edges);
+   }
+
+  void backward(
+      at::TensorList tensors,
+      at::TensorList grad_tensors,
+      c10::optional<bool> keep_graph,
+      bool create_graph) {
```

Comment:
It would be great to preserve the Python equivalent function's signature `torch.autograd.backward(tensors, grad_tensors=None, retain_graph=None, create_graph=False, grad_variables=None)`, so that when a user tries to port code using `torch.autograd.backward` from Python to C++ API it would be easier to do so.

Reply:
I checked the autograd python signature, `grad_variables` is deprecated and not encouraged to use, user will receive runtime error if they specify `grad_variables`. `retain_graph` is the same as `keep_graph`, as we don't have `NamedType` in C++, the user should not notice any difference?

Reply:
Ignoring `grad_variables` looks good. For `retain_graph`, it seems that if its value is defined, it can be different from `create_graph`? https://github.com/pytorch/pytorch/blob/1df57c943f7bbdec8285c9675dbc3093788b783d/torch/autograd/__init__.py#L88-L89 Would be great to verify this